storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +65,387 @@
     }
 
-    private void maybeSyncCommittedDataAndOffsets(boolean forceSync) {
-        // Return immediately if there is no consumption from last time.
-        boolean noConsumedOffsetUpdates = partitionToConsumedOffsets.equals(lastSyncedPartitionToConsumedOffsets);
-        if (noConsumedOffsetUpdates || !forceSync && time.milliseconds() - lastSyncedTimeMs < committedOffsetSyncIntervalMs) {
-            log.debug("Skip syncing committed offsets, noConsumedOffsetUpdates: {}, forceSync: {}", noConsumedOffsetUpdates, forceSync);
+    private void maybeMarkUserPartitionsAsReady() {
+        if (isAllUserTopicPartitionsInitialized) {
             return;
         }
-
-        try {
-            // Need to take lock on assignPartitionsLock as assignedTopicPartitions might
-            // get updated by other threads.
-            synchronized (assignPartitionsLock) {
-                for (TopicIdPartition topicIdPartition : assignedTopicPartitions) {
-                    int metadataPartition = topicPartitioner.metadataPartition(topicIdPartition);
-                    Long offset = partitionToConsumedOffsets.get(metadataPartition);
-                    if (offset != null) {
-                        remotePartitionMetadataEventHandler.syncLogMetadataSnapshot(topicIdPartition, metadataPartition, offset);
-                    } else {
-                        log.debug("Skipping sync-up of the remote-log-metadata-file for partition: [{}] , with remote log metadata partition{}, and no offset",
-                                topicIdPartition, metadataPartition);
+        maybeFetchBeginAndEndOffsets();
+        boolean isAllInitialized = true;
+        for (final UserTopicIdPartition utp : assignedUserTopicIdPartitions.values()) {
+            if (utp.isAssigned && !utp.isInitialized) {
+                final Integer metadataPartition = utp.metadataPartition;
+                final BeginAndEndOffsetHolder holder = offsetHolderByMetadataPartition.get(toRemoteLogPartition(metadataPartition));
+                // The offset-holder can be null, when the recent assignment wasn't picked up by the consumer.
+                if (holder != null) {
+                    final Long readOffset = readOffsetsByMetadataPartition.getOrDefault(metadataPartition, -1L);
+                    // 1) The end-offset was fetched only once during reassignment. The metadata-partition can receive
+                    // new stream of records, so the consumer can read records more than the last-fetched end-offset.
+                    // 2) When the internal topic becomes empty due to breach by size/time/start-offset, then there
+                    // are no records to read.
+                    if (readOffset + 1 >= holder.endOffset || holder.endOffset.equals(holder.beginOffset)) {
+                        markInitialized(utp);
                     }
                 }
```

Comment:
Some debug logs should be added here for troubleshooting, ex: why the topic partition never get initialized?
So, I suggest:
1. log debug when holder is null
2. log debug when holder is not null, but it still cannot mark as initialized, ex: `cannot mark initialized since the readOffset is {}, but endoffset is {}.`

Reply:
Added